src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -951,6 +951,16 @@ void c_typecheck_baset::typecheck_expr_sizeof(exprt &expr)
   }
   else
   {
+    if(
+      type.id() == ID_struct_tag &&
+      to_struct_type(follow_tag(to_struct_tag_type(type))).is_incomplete())
+    {
+      error().source_location = expr.source_location();
+      error() << "calling sizeof for incomplete type: " << to_string(type)
```

Comment:
Looks good in principal, I just have one suggestion, and one query:
* I think you could pretty much copy the compiler error message from gcc for the error message here:
```
gcc cbmc-test.c
cbmc-test.c:8:31: error: invalid application of 'sizeof' to an incomplete type
      'struct foo'
  struct foo* thefoo = malloc(sizeof(struct foo));
```
* This works for incomplete *structs* - but it won't catch other incomplete types? (e.g. unions, arrays with no dimensions, etc?)

Reply:
I've change the message and extended the check to unions and enums. Arrays are weird: at the time of typechecking we already see variables of type `int[]` as if having type `int`.

Reply:
Hmm, so does that mean you don't even know they are an array at that point? Also, what's the difference at that point between a variable of type `int[]` and a variable of type `int[10]` - does one have a length attribute somewhere? If you can know it's an array, and know it's length then it's a complete type. If you don't know it's length, then it's incomplete.

Reply:
My bad: I was looking at a different `sizeof`. Now it should work for enums and arrays as well.